contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql

```diff
@@ -437,93 +437,56 @@ LANGUAGE plpgsql;
-               END
-        INTO column_data_type, column_precision, column_length
-        FROM pg_attribute a
-        WHERE a.attrelid = object_id AND a.attnum = column_id;
-
-        -- Remove delimiters
-        column_data_type := sys.babelfish_remove_delimiter_pair(column_data_type);
-
-        IF column_data_type IS NOT NULL THEN
-            column_length := CASE
-                -- Columns declared with max specifier case
-                WHEN column_length = -1 AND column_data_type IN ('varchar', 'nvarchar', 'varbinary')
-                THEN -1
-                WHEN column_data_type = 'xml'
-                THEN -1
-                WHEN column_data_type IN ('tinyint', 'bit') 
-                THEN 1
-                WHEN column_data_type = 'smallint'
-                THEN 2
-                WHEN column_data_type = 'date'
-                THEN 3
-                WHEN column_data_type IN ('int', 'integer', 'real', 'smalldatetime', 'smallmoney') 
-                THEN 4
-                WHEN column_data_type IN ('time', 'time without time zone')
-                THEN 5
-                WHEN column_data_type IN ('double precision', 'bigint', 'datetime', 'datetime2', 'money') 
-                THEN 8
-                WHEN column_data_type = 'datetimeoffset'
-                THEN 10
-                WHEN column_data_type IN ('uniqueidentifier', 'text', 'image', 'ntext')
-                THEN 16
-                WHEN column_data_type = 'sysname'
-                THEN 256
-                WHEN column_data_type = 'sql_variant'
-                THEN 8016
-                WHEN column_data_type IN ('bpchar', 'char', 'varchar', 'binary', 'varbinary') 
-                THEN column_length
-                WHEN column_data_type IN ('nchar', 'nvarchar') 
-                THEN column_length * 2
-                WHEN column_data_type IN ('numeric', 'decimal')
-                THEN 
-                    CASE
-                        WHEN column_precision IS NULL 
-                        THEN NULL
-                        ELSE ((column_precision + 8) / 9 * 4 + 1)
-                    END
-                ELSE NULL
-            END;
+        -- Check if it's a user-defined data type
+        SELECT sys.translate_pg_type_to_tsql(typbasetype), typlen, typtypmod 
+        INTO column_data_type, typelen, typemod
+        FROM pg_type
+        WHERE oid = (SELECT atttypid FROM pg_attribute WHERE attrelid = object_id AND attnum = column_id);
+
+        IF column_data_type = 'sysname' THEN
+            column_length := 256;
+        ELSE 
+            -- Calculate column length based on base type information
+            column_length := COALESCE(sys.tsql_type_max_length_helper(column_data_type, typelen, typemod), NULL);
         END IF;
+    ELSE
+        column_length := COALESCE(sys.tsql_type_max_length_helper(column_data_type, typelen, typemod), NULL);
+    END IF;
 
-        RETURN column_length::SMALLINT;
-    END;
+    RETURN column_length::SMALLINT;
```

Comment:
Question: Do we really need this CAST to smallint here or should we declare the `column_length` to smallint in the first place?